From 5d7969b1005b2c688057097a46a0a63860d7367c Mon Sep 17 00:00:00 2001
From: Florian Franzen <Florian.Franzen@gmail.com>
Date: Thu, 1 Oct 2026 00:43:08 +0200
Subject: [PATCH] [clr] Do not define __noinline__ for host compilers

host_defines.h defines __noinline__ as an empty macro for non-HIP
compilers. It is a reserved identifier that libstdc++ 16.2 uses as
[[__gnu__::__noinline__]] in <format>, which the macro turns into the
invalid [[__gnu__::]]. Drop the host definition, as CUDA's
host_defines.h does; HIP-compiled code is unaffected.

Issue #9897.
---
 projects/clr/hipamd/include/hip/amd_detail/host_defines.h | 5 ++++-
 projects/hip/docs/how-to/hip_cpp_language_extensions.rst  | 4 ++++
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/projects/clr/hipamd/include/hip/amd_detail/host_defines.h b/projects/clr/hipamd/include/hip/amd_detail/host_defines.h
index aa343848460a..e132518e650e 100644
--- a/hipamd/include/hip/amd_detail/host_defines.h
+++ b/hipamd/include/hip/amd_detail/host_defines.h
@@ -413,7 +413,10 @@ typedef __hip_internal::int64_t __hip_int64_t;
 
 #define __global__
 
-#define __noinline__
+// __noinline__ is deliberately not defined for host compilers. It is a
+// reserved identifier that standard libraries use as an attribute name, e.g.
+// [[__gnu__::__noinline__]] in libstdc++, which an empty macro turns into
+// invalid code.
 #define __forceinline__ inline
 
 #define __shared__
